### PR TITLE
add new script to add 'javaType' property to all schemas, for compatibility with jsonschema2pojo

### PR DESCRIPTION
**Pull request by jacobyavis**
#### What type of PR is this?
/kind feature

#### What this PR does / why we need it:
This adds jsonschema2pojo compatibility as detailed in issue #413

### CODE REVIEW

**Comment by pjohnmeyer** [comment]
I'm not a huge fan of this; based on the issue it seems like we're fixing a bug in `jsonschema2pojo` by changing our schema files.

It does make me think that maybe we should move our "raw" schema definitions into another set of files entirely, then generate the real schema files with all of the "always" stuff like $schema, $id, $ref, $comment, additionalProperties: false ...

**Comment by JSv4** [comment]
You're thinking some kind of scaffolding tool, basically, where we auto-generate new schemas rather than incorporate the properties via $ref?

**Comment by jacobyavis** [comment]
I think this is a good idea. To introduce almost like a 'build' step that transforms the raw schema src into the finished schema.

So given a src file like so: 
```{
  "$id": "https://raw.githubusercontent.com/Open-Cap-Table-Coalition/Open-Cap-Format-OCF/main/schema/types/Email.schema.json",
  "title": "Type - Email",
  "description": "Type representation of an email address",
  "type": "object",
  "properties": {
    "email_type": {
      "description": "Type of e-mail address (e.g. personal or business)",
      "$ref": "https://raw.githubusercontent.com/Open-Cap-Table-Coalition/Open-Cap-Format-OCF/main/schema/enums/EmailType.schema.json"
    },
    "email_address": {
      "description": "A valid e-mail address",
      "type": "string",
      "format": "email"
    }
  },
  "required": ["email_type", "email_address"]
}
```

The build step would generate a finished schema like so:
``` {
  "$schema": "http://json-schema.org/draft-07/schema",
  "$id": "https://raw.githubusercontent.com/Open-Cap-Table-Coalition/Open-Cap-Format-OCF/main/schema/types/Email.schema.json",
  "title": "Type - Email",
  "description": "Type representation of an email address",
  "type": "object",
  "properties": {
    "email_type": {
      "description": "Type of e-mail address (e.g. personal or business)",
      "$ref": "https://raw.githubusercontent.com/Open-Cap-Table-Coalition/Open-Cap-Format-OCF/main/schema/enums/EmailType.schema.json"
    },
    "email_address": {
      "description": "A valid e-mail address",
      "type": "string",
      "format": "email"
    }
  },
  "required": ["email_type", "email_address"],
  "additionalProperties": false,
  "$comment": "Copyright © 2023 Open Cap Table Coalition (https://opencaptablecoalition.com) / Original File: https://github.com/Open-Cap-Table-Coalition/Open-Cap-Format-OCF/tree/main/schema/types/Email.schema.json",
  "javaType": "com.opencaptablecoalition.ocf.schema.types.Email"
}
```

One benefit this brings us is being able to ignore the generated sources, so PRs like this one (and many others, are not difficult to review, 146 schema files!)

I'd be interested in tackling this as a separate issue, but I'm still hoping to get `javaType` in as is for the next release, so that the attribute will appear on the hosted schemas at `raw.githubusercontent.com`

**Comment by jacobyavis** [comment]
To expand on the idea a bit further, it would be nice if the generated schemas which use Composition had the properties expanded out further.

Currently jsonschema2pojo treats anything listed with empty braces as an unknown type, generic `Object`, so I already am planning to tackle this in some way. Perhaps this makes the most sense to do all together

File: utils/schema-utils/AddJavaTypeToSchemas.ts [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+ * @param schema_path -> Local path where schema_obj was loaded from. WILL BE OVERWRITTEN with serialized data from updated schema_obj.
+ * @param schema_inst -> OCF schema JSON to update.
+ * @param verbose -> Boolean: Display detailed logs.
+ */
+export function addJavaTypePropertyToSchema(
+  schema_path: string,
+  schema_inst: Record<string, any>,
+  verbose: boolean = false
+) {
+  let java_type_contents = `com.opencaptablecoalition.ocf.schema.${basenameRelativePathToSchemaDir(
```
Comment (jacobyavis) [inline utils/schema-utils/AddJavaTypeToSchemas.ts:21]: I've added `opencaptablecoalition` as domain here, thanks for the pointer @pjohnmeyer

**Comment by pjohnmeyer** [comment]
> Note that I think $id should be included in raw since it's used in the various $refs

I wouldn't, I'd introduce some kind of short ID to our raw files for refs with some kind of wart like:

`$ref: $enums/EmailType.schema.json`

> it would be nice if the generated schemas which use Composition had the inherited properties exploded with full details

Hmmmm maybe? I'm less sold on this.

I'd be willing to take a stab at this after we get the next release out.

**Comment by jacobyavis** [comment]
> I wouldn't, I'd introduce some kind of short ID to our raw files for refs with some kind of wart like:
> 
> `$ref: $enums/EmailType.schema.json`

Ah yes, duh. I wasn't thinking clearly.

> > it would be nice if the generated schemas which use Composition had the inherited properties exploded with full details

> Hmmmm maybe? I'm less sold on this.

Having these properties exploded would most likely help the schema to behave properly with various JSON Schema tools out there (besides just jsonschema2pojo, which I'm focused on right now), because our composition pattern using `allOf` is non-standard. 

For example, think how much of an easier time we might have had writing the doc generator if the schema properties were exploded. But obviously for development, we want the raw src to be simplified/pure, no chance for unintentional overrides

**Comment by JSv4** [comment]
> > I wouldn't, I'd introduce some kind of short ID to our raw files for refs with some kind of wart like:
> > `$ref: $enums/EmailType.schema.json`
> 
> Ah yes, duh. I wasn't thinking clearly.
> 
> > > it would be nice if the generated schemas which use Composition had the inherited properties exploded with full details
> 
> > Hmmmm maybe? I'm less sold on this.
> 
> Having these properties exploded would most likely help the schema to behave properly with various JSON Schema tools out there (besides just jsonschema2pojo, which I'm focused on right now), because our composition pattern using `allOf` is non-standard.
> 
> For example, think how much of an easier time we might have had writing the doc generator if the schema properties were exploded. But obviously for development, we want the raw src to be simplified/pure, no chance for unintentional overrides

I like this too. I have some python code that does most of this for my own purposes. Happy to contribute or refactor in TS.

**Comment by jacobyavis** [comment]
@JSv4 @pjohnmeyer Can we merge this in the meantime? 
For us to move forward I'd like the javaType to be available at the remote refs by the time we cut the 1.1.0 release

**Comment by regebro** [comment]
I'm not sure why this would be necessary, the information is not actually new information, it's just the location of the schema file, transformed into a Java object name in a simple transformation. That seems to me like it could be a part of the tool that generates the java types.

**Comment by JSv4** [comment]
@jacobyavis, could this just be part of the loader script in whatever tooling you're working on?

**Comment by jacobyavis** [comment]
For sure, it could be just part of the tool that loads the java types. I definitely debated keeping this whole thing from the TWG, but there's two main reasons I decided to create this PR.

1. If there's another corp to come along and decide that jsonschema2pojo is the simplest way forward for them, they will likely run into the same issue/bug that I did. If this change goes in, they would find that integration much more seamless.

2. As I have it locally today, I need processing to change all $refs to local references because the actual remote refs do not carry the `javaType` definition that the lib is looking for

**Comment by jacobyavis** [comment]
After discussion in the TWG meeting yesterday, going to look into creating a separate repo, something like ocf4java which does the required schema manipulations for full jsonschema2pojo compatibility, then deliver the gen'd files as a jar.

**Pull request closed without merging**